### PR TITLE
List corpuses

### PR DESCRIPTION
**Pull request by blakerosenthal**
Addresses https://github.com/Quansight/ragna/issues/492

### CODE REVIEW

**Comment by blakerosenthal** [comment]
@pmeier Is this about right? I like the idea of raising `NotImplemented` on `source_storage.list_corpuses()` in case the user decides to ignore corpuses altogether.

Also right now the `/corpuses` endpoint returns the list of corpus names directly from the source_storages. I have two questions about this:
1. Should the api call distinguish which source storage the corpus name belongs to? E.g. `["LanceDB/<corpus_name>", "ChromaDB/<corpus_name>"]`? Maybe actually returning a JSON of `{"LanceDB": [<corpus_name>], "ChromaDB": [<corpus_name>]}` would be better?
2. For the built in source storages, the default corpus name is just the embedding ID. Maybe `list_corpuses()` should translate this to a prettier string like `"default"` for the sake of the UI.

Finally, should I add the actual UI changes as part of this issue, or will that be a separate issue?

**Comment by pmeier** [comment]
> I like the idea of raising `NotImplemented` on `source_storage.list_corpuses()` in case the user decides to ignore corpuses altogether.

On second thought, we should probably raise a descriptive `RagnaException` instead. We need to tell users what they have to do if they encounter this error and then exit gracefully.

> 1. Should the api call distinguish which source storage the corpus name belongs to? E.g. `["LanceDB/<corpus_name>", "ChromaDB/<corpus_name>"]`? Maybe actually returning a JSON of `{"LanceDB": [<corpus_name>], "ChromaDB": [<corpus_name>]}` would be better?

You are right, that needs to happen. I would go for the JSON approach where the key is the `.display_name()`.


> 2\. For the built in source storages, the default corpus name is just the embedding ID. Maybe `list_corpuses()` should translate this to a prettier string like `"default"` for the sake of the UI.

Yes, that needs to happen on the source storage itself. I'm also inclined to remove the feature of using the embedding ID as name for now. There is no use and it makes it more complicated. So feel free to just remove the embedding ID and just use `"default"`.

**Comment by pmeier** [comment]
> Finally, should I add the actual UI changes as part of this issue, or will that be a separate issue?

Up to you. When using two PRs we can merge this PR soon and don't have to wait for #484 to be merged.

**Review by pmeier — COMMENTED** [review]
Looking good. I'll review properly when the open questions you had are addressed.

**Comment by blakerosenthal** [comment]
> Up to you. When using two PRs we can merge this PR soon and don't have to wait for #484 to be merged.

Okay, let's keep them separate then

File: tests/deploy/api/test_e2e.py [hunk 1]
```diff
@@ -86,6 +89,9 @@ def test_e2e(tmp_local_root, multiple_answer_chunks, stream_answer, corpus_name)
         assert len(chat["messages"]) == 1
         assert chat["messages"][-1] == message
 
+        corpuses = client.get("/corpuses").raise_for_status().json()
+        assert corpuses == {source_storage: [str(corpus_name)]}
```
Comment (pmeier) [inline tests/deploy/api/test_e2e.py:93]:
```suggestion
        assert corpuses == {source_storage: [corpus_name]}
```

File: ragna/deploy/_api/core.py [hunk 2]
```diff
@@ -145,6 +145,19 @@ async def get_components(_: UserDependency) -> schemas.Components:
             ],
         )
 
+    @app.get("/corpuses")
+    async def get_corpuses(_: UserDependency) -> dict[str, list[str]]:
```
Comment (pmeier) [inline ragna/deploy/_api/core.py:149]: Following the discussion in https://github.com/Quansight/ragna/issues/493#issuecomment-2301776184 should we introduce a `source_storage` query parameter?

```suggestion
    async def get_corpuses(_: UserDependency, source_storage: Optional[str] = None) -> dict[str, list[str]]:
```

And with that we can do

```python
if source_storage is not None:
    source_storages = [components_map[source_storage]]
else:
    ...
```

Meaning, if the user does `GET /corpuses` they get everything, but they can do `GET /corpuses?source_storage=Chroma` to pre-filter it. Still, the output format should be the same in both cases.

**Review by pmeier — COMMENTED** [review]
Thanks Blake, almost LGTM to me!

File: ragna/deploy/_api/core.py (hunk 2 above)
Comment (blakerosenthal) [reply ragna/deploy/_api/core.py:149]: I like that. I'll push that up.

File: tests/deploy/api/test_e2e.py [hunk 3]
```diff
@@ -86,6 +90,21 @@ def test_e2e(tmp_local_root, multiple_answer_chunks, stream_answer, corpus_name)
         assert len(chat["messages"]) == 1
         assert chat["messages"][-1] == message
 
+        corpuses = client.get("/corpuses").raise_for_status().json()
+        assert corpuses == {source_storage: [corpus_name]}
+
+        corpuses = (
+            client.get(f"/corpuses?source_storage={source_storage}")
```
Comment (pmeier) [inline tests/deploy/api/test_e2e.py:97]: Let the library handle the URL building.

```suggestion
            client.get("/corpuses", params={"source_storage": source_storage})
```

Same for the occurrence below.

**Review by pmeier — APPROVED** [review]
Thanks Blake. FYI: I pushed a small cleanup commit.

**Pull request merged**